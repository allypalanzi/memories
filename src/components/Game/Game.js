import React, { Component } from 'react'

import Timer from '../Timer/Timer'
import Card from '../Card/Card'
import Button from '../Button/Button'
import styles from './Game.scss'

class Game extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      level: 'easy',
      cards: [],
      flipped: 0,
      lastCard: {},
      matches: 0,
      maxMatches: 0,
    }
  }
  async getCards() {
    try {
      const response = await fetch('https://web-code-test-dot-nyt-games-prd.appspot.com/cards.json');
      const cards = await response.json();
      const leveledCards = cards.levels.find(card => card.difficulty === this.state.level);
      const cardsState = leveledCards.cards.map((card, index) => {
        return { value: card, matched: false, flipped: false, id: index }
      });
      this.setState({cards: cardsState, loading: false, maxMatches: cardsState.length / 2})
    } catch (e) {
      console.error(e)
    }
  };


  handleClick(value, id) {
    const cards = this.state.cards;
    const selectedCard = cards[id];
    const lastCard = this.state.lastCard;
    let matches = this.state.matches;
    let flipped = this.state.flipped;
    if (lastCard.id !== selectedCard.id && lastCard.value === selectedCard.value) {
      selectedCard.matched = true;
      cards[lastCard.id].matched = true;
      matches++
      flipped = 0;
    }
    if (flipped === 2) {
      cards.forEach(card => card.flipped = false);
      flipped = 0;
    }
    selectedCard.flipped = !selectedCard.flipped;
    const flippedValue = selectedCard.flipped ? flipped + 1 : flipped - 1;
    this.setState({ cards, flipped: flippedValue, lastCard: selectedCard, matches: matches });
  }

  handleReset() {
    const cards = this.state.cards;
    cards.forEach(card => {
      card.flipped = false;
      card.matched = false;
    });
    this.setState({ cards, flipped: 0, maches: 0, lastCard: {} });
  }

  toggleLevel() {
    const level = this.state.level === 'easy' ? 'hard' : 'easy';
    this.setState({ level });
    this.getCards();
  }

  componentWillMount() {
    this.getCards();
  }

  render() {
    const { cards, loading, flipped, maxMatches, matches, level } = this.state;
    if (loading) {
      return (<div>loading</div>)
    }
    const cardDisplay = cards.map((card, index) => <Card onClick={e => this.handleClick(card.value, card.id)} flipped={card.flipped} key={index} id={card.id} value={card.value} matched={card.matched} />);
    return (
      <div className={styles.wrapper}>
        {matches === maxMatches && <h1>u win</h1>}
        <div className={styles.cards}>
          {cardDisplay}
        </div>
        <Button onClick={e => this.handleReset()} text={'Reset'} />
        <Button onClick={e => this.toggleLevel()} text={`Change level to ${level === 'easy' ? 'hard' : 'easy'}`} />
        <Timer />
      </div>
    )
  };
};

export default Game
