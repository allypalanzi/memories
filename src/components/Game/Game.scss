.cards {
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  align-items: space-around;
  text-align: center;
  justify-content: space-between;
  padding: 2em 0;
}

.utilites {
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin: 1em;
}
