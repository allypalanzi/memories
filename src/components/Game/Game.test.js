import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import Game from './Game'

jest.mock('../../client/services')

describe('<Game />', () => {
  const renderedComponent = shallow(<Game />)

  it('renders correctly', () => {
    renderedComponent.update()
    expect(toJson(renderedComponent)).toMatchSnapshot()
  })

  it('When handleReset is clicked, flipped and matched cards are set to false', () => {
    const cards = renderedComponent.state().cards
    // Set a card to flipped
    cards[0].flipped = true
    renderedComponent.setState({ cards })
    // Call reset
    renderedComponent.find('#qa-reset').simulate('click')
    const resetCards = renderedComponent.state().cards.find(card => card.flipped || card.matched)
    expect(resetCards).toBeUndefined()
  })

  it('Component begins on level easy, then when toggle is clicked, it updates to hard', () => {
    expect(renderedComponent.state().level).toBe('easy')
    renderedComponent.find('#qa-toggle').simulate('click')
    expect(renderedComponent.state().level).toBe('hard')
  })
})
