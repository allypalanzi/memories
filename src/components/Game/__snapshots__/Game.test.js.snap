// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> renders correctly 1`] = `
<div
  className="game"
>
  <div
    className="cards"
  >
    <Card
      disabled={false}
      flipped={false}
      id={0}
      key="0_✈"
      matched={false}
      onClick={[Function]}
      value="✈"
    />
    <Card
      disabled={false}
      flipped={false}
      id={1}
      key="1_♘"
      matched={false}
      onClick={[Function]}
      value="♘"
    />
    <Card
      disabled={false}
      flipped={false}
      id={2}
      key="2_✈"
      matched={false}
      onClick={[Function]}
      value="✈"
    />
    <Card
      disabled={false}
      flipped={false}
      id={3}
      key="3_♫"
      matched={false}
      onClick={[Function]}
      value="♫"
    />
    <Card
      disabled={false}
      flipped={false}
      id={4}
      key="4_♫"
      matched={false}
      onClick={[Function]}
      value="♫"
    />
    <Card
      disabled={false}
      flipped={false}
      id={5}
      key="5_☆"
      matched={false}
      onClick={[Function]}
      value="☆"
    />
    <Card
      disabled={false}
      flipped={false}
      id={6}
      key="6_♘"
      matched={false}
      onClick={[Function]}
      value="♘"
    />
    <Card
      disabled={false}
      flipped={false}
      id={7}
      key="7_☆"
      matched={false}
      onClick={[Function]}
      value="☆"
    />
  </div>
  <div
    className="utilites"
  >
    <div>
      <Button
        id="qa-reset"
        onClick={[Function]}
        text="Reset"
      />
      <Button
        id="qa-toggle"
        onClick={[Function]}
        text="Change level to hard"
      />
    </div>
  </div>
</div>
`;
